### ISSUE TITLE
Idea: Exploit hierarchical composition in simulation generation

### ISSUE THREAD

**Issue by lukem12345** [AlgebraicJulia/Decapodes.jl#266]
Currently, we generate simulation code only on single Decapode objects, after oapply is performed.

However, we could take advantage of repeated sub components which only differ in the pointers to their inputs and outputs. For each node in a composition diagram for which the same Decapode is plugged in, we could compile it to a simulation helper function, which could be re-used inside the main ODEProblem function generated.

This will surely result in simulation functions with fewer total lines of code. However, it is not apparent whether this would result in more efficient generated code that the Julia compiler can optimize, or whether this could improve pre-compile times.

So, we should manually create an example of this generated code and compare. A reasonable example would use the unconnected RelationDiagram of 2 or more distinct nodes, plugging in e.g. the Brusselator Decapode for each component. (I.e. The strategy used for [non-interacting multi-species Navier-Stokes](https://github.com/AlgebraicJulia/Decapodes.jl/issues/70#issuecomment-1421598346).)

**Comment by jpfairbanks** [comment]
MTK also generates code but at the scalar level. I think they generate basic blocks with on the order of 100k lines. I don't think that we are pushing the boundaries of code generation in Julia with even our biggest decapodes.

**Comment by lukem12345** [comment]
Yeah I'm wondering where the cutoff is in terms of effectively inlining with copies of the same Decapode.
